usr/src/common/zfs/zfs_namecheck.c

```diff
@@ -183,7 +184,8 @@ dataset_namecheck(const char *path, namecheck_err_t *why, char *what)
 				 * Make sure this is really a zero-length
 				 * component and not a '@@'.
 				 */
-				if (*end == '@' && found_snapshot) {
+				if ((*end == '@' || *end == '#') &&
+				    found_separator) {
 					*why = NAME_ERR_MULTIPLE_AT;
```

Comment:
Technically, this is no longer AT only.  Maybe we should rename NAME_ERR_MULTIPLE_AT to something more generic, like NAME_ERR_MULTIPLE_SEPARATOR.